### PR TITLE
Post: Add comment moderation tools

### PR DESCRIPTION
**Pull request by rodrigoi**
Adds moderation tools to the comments under Posts. Works together with https://github.com/Automattic/wp-calypso/pull/12295.

This is hidden under the `comments/moderation-tools-in-posts` feature flag.

fireshot capture 1 - blog posts free site wordpress com_ - http___calypso localhost_3000_posts

fireshot capture 3 - blog posts free site wordpress com_ - http___calypso localhost_3000_posts

fireshot capture 2 - blog posts free site wordpress com_ - http___calypso localhost_3000_posts

Part of #11209.

### CODE REVIEW

File: client/blocks/comments/comment-approve-action.jsx [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+const CommentApproveAction = ( { translate, status, onClick } ) => {
+	const isApproved = status === 'approved';
+	const buttonStyle = classnames( 'comments__comment-actions-approve', {
+		'is-approved': isApproved
+	} );
+
+	return (
+		<button className={ buttonStyle } onClick={ onClick }>
+			<Gridicon icon="checkmark" size={ 18 }/>
+			<span className="comments__comment-actions-like-label">{ isApproved ? translate( 'Approved' ) : translate( 'Approve' ) }</span>
```
Comment (a8ci18n) [inline client/blocks/comments/comment-approve-action.jsx]: Hi! I've found a possible matching string that has already been translated **31** times:
`translate( 'Approved', { context: 'comment status'} )` *ES Score: 20*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/comment-moderation-tools-on-post)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/blocks/comments/comment-approve-action.jsx [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+const CommentApproveAction = ( { translate, status, approveComment, unapproveComment } ) => {
+	const isApproved = status === 'approved';
+	const buttonStyle = classnames( 'comments__comment-actions-approve', {
+		'is-approved': isApproved
+	} );
+
+	return (
+		<button className={ buttonStyle } onClick={ ! isApproved ? approveComment : unapproveComment }>
+			<Gridicon icon="checkmark" size={ 18 } />
+			<span className="comments__comment-actions-like-label">{ isApproved ? translate( 'Approved' ) : translate( 'Approve' ) }</span>
```
Comment (a8ci18n) [inline client/blocks/comments/comment-approve-action.jsx]: :disappointed: 31 existing translations will be lost with this change.
If the string must change, the following option has already been translated **28** times:
`translate( 'Approved', { context: 'verb: past-tense'} )` *ES Score: 18*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by roundhill** [comment]
If I edit a comment, I see `<p>` tags around it, which I don't see around the comment in wp-admin:

screen shot 2017-03-20 at 3 48 29 pm

screen shot 2017-03-20 at 3 49 22 pm

**Comment by roundhill** [comment]
Another weird thing when editing a comment, the edit UI closes after I click `Send`. Also, the `Send` button should probably be `Save` or `Edit`?

edit comment close

**Comment by roundhill** [comment]
Since `Trash` is destructive, I'd move it over to the right side, and move edit to before Spam. So:

`Approve/Unapprove` `Edit` `Spam` `Trash`

**Comment by roundhill** [comment]
> Another weird thing when editing a comment, the edit UI closes after I click Send
Actually this happens when you make any moderation action. Maybe we could disable the comment you are editing and show a spinner while the action is taking place?

File: client/blocks/comments/comment-actions.jsx [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+					<span className="comments__comment-actions-like-label">{ translate( 'Spam' ) }</span>
+				</button>
+			}
+			{ isEnabled( 'comments/moderation-tools-in-posts' ) &&
+				<button className="comments__comment-actions-edit" onClick={ editComment }>
+					<Gridicon icon="pencil" size={ 18 } />
+					<span className="comments__comment-actions-like-label">{ translate( 'Edit' ) }</span>
+				</button>
+			}
+			{ isEnabled( 'comments/moderation-tools-in-posts' ) &&
```
Comment (roundhill) [inline client/blocks/comments/comment-actions.jsx]: I think you'll need to add some checks for the user type. Certain WordPress user roles can't moderate comments, so we'd want to hide the moderate actions for them.

https://codex.wordpress.org/Roles_and_Capabilities

File: client/blocks/comments/comment-actions.jsx [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+				<button className="comments__comment-actions-reply" onClick={ handleReply }>
+					<Gridicon icon="reply" size={ 18 } />
+					<span className="comments__comment-actions-reply-label">{ translate( 'Reply' ) }</span>
+				</button>
+			}
+			{ showCancelReplyButton &&
+				<button className="comments__comment-actions-cancel-reply" onClick={ onReplyCancel }>{ translate( 'Cancel reply' ) }</button>
+			}
+			{ showCancelEditButton &&
+				<button className="comments__comment-actions-cancel-reply" onClick={ editCommentCancel }>{ translate( 'Cancel' ) }</button>
```
Comment (roundhill) [inline client/blocks/comments/comment-actions.jsx]: There's a few long lines here, might be good to wrap the attributes of the `<button>`s here.

**Comment by drw158** [comment]
1 Would be nice to get this icon orange:

screen shot 2017-03-21 at 10 47 58 am

2 We should align hover styles with notifications. The hover styles on the actions are orange, but I think we can do better. What about this?

- Orange hover color: Approve, Like
- Red hove color: Spam, Trash
- Blue hover color: Reply, Edit

3 I agree with Dan, should be `Save` in this case.
screen shot 2017-03-21 at 10 48 21 am

4 @roundhill: Not sure about the ordering. I like that the `Cancel` button is right by the Edit button (while in edit mode). The weird order could be solved by my number 5 comment though 👇 

5 I think some work will need to be done with the action buttons. This gets a bit messy (edit a reply, then hit edit)

screen shot 2017-03-21 at 10 57 37 am

I think we should:

- Move the cancel button above the text field, to be consistent with `cancel reply`
- Somehow disable/hide the other actions while editing.
- Rename it `Cancel edit` instead of `Cancel`

So, overall maybe something like this:

screen shot 2017-03-21 at 11 24 20 am

Alternate design:

screen shot 2017-03-21 at 11 31 09 am



6 ~I can't edit other comments, while editing a comment. We could just disable all other Edit buttons while editing, or somehow enable editing multiple comments at once? Along the same lines,  I can't edit a comment while I am replying to one. We should just disable the buttons in these cases.~ For some reason, now I can. Weird!



Also, let's not forget to test this in Horizon when we are ready, I think it will be a good candidate.

**Comment by folletto** [comment]
Good feedback there.

Just a question: is there any reason why we want all the options always visible, instead of:

* If it's not approved yet, show the options 
* If it's already approved, show always •••

The rationale of this suggestion is that the fully opened option add a lot of visual noise on that page, and if I'm not actively moderating, it's unlikely I'll ever need these options.
Thus, these options are valuable and surfaced only when the comment hasn't been flagged yet.

An alternative idea, that I feel could be even better:

comment-mod

In this way we clearly indicate the status of the comment (we can even just not show the ✓) and when it's marked to review then the indicator is clearly visible too.

The ••• would always open the dropdown as the above.

The position can also be on the same line of Reply and Like, just right-aligned.

**Comment by roundhill** [comment]
Let's go ahead and merge this and add the fixes in additional PRs. :shipit:

**Comment by bluefuton** [comment]
@rodrigoi we also use the comments block in Reader full post view, for example:

https://wpcalypso.wordpress.com/read/feeds/7727265/posts/1390361935

So we're currently seeing this:

screen shot 2017-03-24 at 16 51 25

Can we have the comment moderation tools activated by an additional prop please?

**Comment by rodrigoi** [comment]
@bluefuton absolutely. This is under a feature flag so rest assured it won't show to users. I'm still working on the comments block and I still have to enforce the distinction between the posts and the reader 😉

**Comment by bluefuton** [comment]
Thanks @rodrigoi :)

**Comment by alisterscott** [comment]
Reverted in #12542 as this bundled with #12295 which introduced a regression #12540

**Pull request merged**